### ISSUE TITLE
Improve invoke() method to avoid hang problems

### ISSUE THREAD

**Issue by DiegoQueiroz** [pywinauto/pywinauto#1057]
This ia a new issue based in the discussion that started in https://github.com/pywinauto/pywinauto/issues/865

Basically, sometimes `invoke()` is raising a generic exception.

```
>>> loginButton.click()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\diego.queiroz\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.8_qbz5n2kfra8p0\LocalCache\local-packages\Python38\site-packages\pywinauto\controls\uia_controls.py", line 121, in click
    self.invoke()
  File "C:\Users\diego.queiroz\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.8_qbz5n2kfra8p0\LocalCache\local-packages\Python38\site-packages\pywinauto\controls\uiawrapper.py", line 513, in invoke
    self.iface_invoke.Invoke()
_ctypes.COMError: (-2147220991, 'event cannot call any subscribers', (None, None, None, 0, None))
```

Despite the exception, **the button is clicked as expected**.
The current workaround is to just ignore the exception, as follows:

```
try:
  loginButton.click()
except:
  pass # ignore

# script continues here
```

This may be happening because the application is not returning in a timely manner (indeed, the application I am testing freezes for a second when I click the login button). Probably a improvement like a configurable timeout may solve this problem.

Also, it would be interesting to raise a more specific exception when an error like this happens.

@vasily-v-ryabov said:
> Sometimes we also face with hang problem when calling .invoke() (.click() is an alias of .invoke() for ButtonWrapper). It means we need to create a separate thread with some timeout and print a warning if it is timed out. This is also separate issue, but we can combine it in one issue since it is all in .invoke() method.

**Comment by vasily-v-ryabov** [comment]
Well, the problem with `RuntimeWarning` is that the same error code can be raised without any real action like in #1029. If we suppress such error, other case will suffer. All such cases are very application specific. So I'd rather leave it unchanged. There are too many workarounds already in pywinauto code. Don't want to grow it without serious reason.

I will keep this issue open for `.invoke()` hang case only.

**Comment by DiegoQueiroz** [comment]
Do you understand your comments are somewhat confusing, right? It seems you're talking with yourself, but it hard for a person like me to understand.

I don't know if it is too much to ask for, but it would be great if you write a FAQ or something to describe better these issues you face with "UIA vs. odd application behavior", so more people will understand the problem and maybe be more helpful.

About this problem, I don't think suppressing the error is correct (it is a workaround for my specific case). But something should be adjusted in pywinauto, so the proper exception is raised (or none, if there's nothing to raise).

**Comment by vasily-v-ryabov** [comment]
FAQ for UIA related application issues is a good idea. It can be added to #929.

I don't think it should be adjusted in pywinauto. I see potential improvement in `comtypes` only: `COMError` shows error code as signed int, but it's much more useful to show unsigned int in hex. Because Microsoft's [COM error codes document](https://docs.microsoft.com/en-us/windows/win32/com/com-error-codes) lists all error codes as hex. For example, dive into Generic error codes. Your case error code is 0x80040201: `EVENT_E_ALL_SUBSCRIBERS_FAILED`. Also these constants should be declared somewhere in `comtypes`. Maybe pywinauto docs should point it.

Declaring additional exception class in pywinauto doesn't make much sense in my mind if we document `comtypes` errors properly.

**Comment by celsolarussa** [comment]
@DiegoQueiroz I have the same problem, could you tell me if you found any solution to use .click()?

**Comment by DiegoQueiroz** [comment]
@celsolarussa The current workaround is to just ignore the exception, as shown in the first message.

```
try:
  loginButton.click()
except:
  pass # ignore exception

# script continues here
```

**Comment by vasily-v-ryabov** [comment]
It is implemented in `atspi` branch. Thanks @eltimen !
Guys, please try this branch:
```
pip uninstall -y pywinauto
pip install https://github.com/pywinauto/pywinauto/archive/refs/heads/atspi.zip
```

**Comment by darkworks** [comment]
> It is implemented in `atspi` branch. Thanks @eltimen ! Guys, please try this branch:
> 
> ```
> pip uninstall -y pywinauto
> pip install https://github.com/pywinauto/pywinauto/archive/refs/heads/atspi.zip
> ```

this branch have no documentation , so be careful if you are on released 0.6.8 and update to this un released branch it will break your script it a lot of places like `child windows` are replaced by `by()` and other methods like `wait_not('visible',timeout = 60)` will also not work. 

am not python pro however i not understand this patch is it even tested ?? because i copied this invoke defination to old 0.6.8 released branch and it give me error ` Error! <class 'NameError'> occurred. `

```py
    def invoke(self):
        """An interface to the Invoke method of the Invoke control pattern"""
        name = self.element_info.name
        control_type = self.element_info.control_type
        invoke_pattern_iface = self.iface_invoke

        # sometimes .Invoke() can hang, although an action (such as button click) was completed successfully
        def watchdog():
            thread = threading.Thread(target=invoke_pattern_iface.Invoke)
            thread.daemon = True
            thread.start()
            thread.join(2.)
            if thread.isAlive():
                warnings.warn('Timeout for InvokePattern.Invoke() call was exceeded', RuntimeWarning)
        watchdog_thread = threading.Thread(target=watchdog)
        watchdog_thread.start()
        watchdog_thread.join(Timings.after_invoke_wait)

        if name and control_type:
            self.actions.log("Invoked " + control_type.lower() + ' "' + name + '"')
        # Return itself to allow action chaining
        return self
```


`watchdog` is define inside invoke and not called ??

**Comment by vasily-v-ryabov** [comment]
We will try to improve backward compatibility before release. But this particular feature is implemented and I'm closing the issue to help us tracking progress for 0.7.0 milestone. Thanks for understanding! I will notify when new release is out.
